### PR TITLE
Porting DTNN Class

### PR DESCRIPTION
**Pull request by GreatRSingh**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/tests/assets/qm9_mini.sdf.csv [hunk 1]
```diff
@@ -0,0 +1,22 @@
+mol_id,A,B,C,mu,alpha,homo,lumo,gap,r2,zpve,u0,u298,h298,g298,cv,u0_atom,u298_atom,h298_atom,g298_atom
```
Comment (tonydavis629) [inline deepchem/models/tests/assets/qm9_mini.sdf.csv:1]: What is this CSV for? It seems to not be used in your tests

File: deepchem/models/torch_models/dtnn.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    In this class, we establish a sequential model for the Deep Tensor Neural Network (DTNN) [1]_.
+
+    Examples
+    --------
+    >>> import os
+    >>> import torch
+    >>> from deepchem.models.torch_models import dtnn
+    >>> from deepchem.data import SDFLoader
+    >>> from deepchem.feat import CoulombMatrix
+    >>> from deepchem.models.torch_models import dtnn
```
Comment (tonydavis629) [inline deepchem/models/torch_models/dtnn.py:25]: dtnn imported twice

File: deepchem/models/torch_models/dtnn.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    >>> data = loader.create_dataset(dataset_file, shard_size=100)
+    >>> inputs = dtnn._compute_features_on_batch(data.X)
+    >>> atom_number, distance, atom_membership, distance_membership_i, distance_membership_j = inputs
+    >>> inputs = [torch.tensor(atom_number).to(torch.int64),
+    ...           torch.tensor(distance).to(torch.float32),
+    ...           torch.tensor(atom_membership).to(torch.int64),
+    ...           torch.tensor(distance_membership_i).to(torch.int64),
+    ...           torch.tensor(distance_membership_j).to(torch.int64)]
+    >>> n_tasks = data.y.shape[0]
+    >>> model = dtnn.DTNN(n_tasks)
```
Comment (tonydavis629) [inline deepchem/models/torch_models/dtnn.py:40]: this should probably be `model = dtnn(n_tasks)` right?

File: deepchem/models/torch_models/dtnn.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    >>> from deepchem.data import SDFLoader
+    >>> from deepchem.feat import CoulombMatrix
+    >>> from deepchem.models.torch_models import dtnn
+    >>> # Get Data
+    >>> model_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    >>> dataset_file = os.path.join(model_dir, 'tests/assets/qm9_mini.sdf')
+    >>> TASKS = ["alpha", "homo"]
+    >>> loader = SDFLoader(tasks=TASKS, featurizer=CoulombMatrix(29), sanitize=True)
+    >>> data = loader.create_dataset(dataset_file, shard_size=100)
+    >>> inputs = dtnn._compute_features_on_batch(data.X)
```
Comment (tonydavis629) [inline deepchem/models/torch_models/dtnn.py:32]: you should not define `_computer_features_on_batch` to be a private helper function. It should be something like DTNN_Featurizer, like this
`featurizer = DTNN_Featurizer(num_tasks=data.y.shape[0]`
`inputs = featurizer(data.X)`
In this way this featurization can be used by other models in a standardized way.

File: deepchem/models/torch_models/dtnn.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    distance_min: float (default -1)
+        minimum distance of atom pairs (in Angstrom)
+    distance_max: float (default = 18)
+        maximum distance of atom pairs (in Angstrom)
+    n_distance: int (default 100)
+        granularity of distance matrix
+        step size will be (distance_max-distance_min)/n_distance
+
+    Computed Features
+    -----------------
```
Comment (tonydavis629) [inline deepchem/models/torch_models/dtnn.py:171]: should be `Returns`, and the doc should explain that features is a list of these np arrays, not separate values

**Review by tonydavis629 — COMMENTED** [review]
A couple things to modify but mostly there

**Review by shaipranesh2 — APPROVED** [review]

**Review by tonydavis629 — APPROVED** [review]
Just those imports and doc formatting and it's good to go

File: deepchem/models/torch_models/dtnn.py (hunk 5 above)
Comment (GreatRSingh) [reply deepchem/models/torch_models/dtnn.py]: Done

File: deepchem/models/torch_models/dtnn.py (hunk 2 above)
Comment (GreatRSingh) [reply deepchem/models/torch_models/dtnn.py]: Fixed

File: deepchem/models/torch_models/dtnn.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+from deepchem.models.losses import L2Loss
+from deepchem.models.torch_models import layers
+from deepchem.models.torch_models import TorchModel
+from deepchem.data.datasets import Dataset
+
+
+class DTNN(nn.Module):
+    """Deep Tensor Neural Networks
+
+    In this class, we establish a sequential model for the Deep Tensor Neural Network (DTNN) [1]_.
```
Comment (rbharath) [inline deepchem/models/torch_models/dtnn.py:16]: Can you add more docs here about DTNN?

File: deepchem/models/torch_models/dtnn.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+        dtnn_step = nn.Dropout(self.dropout)(dtnn_embedding)
+        dtnn_gather = self.dtnn_gather([dtnn_step, inputs[2]])
+
+        dtnn_gather = nn.Dropout(self.dropout)(dtnn_gather)
+        output = self.linear(dtnn_gather)
+        return output
+
+
+def _compute_features_on_batch(X_b: np.ndarray,
```
Comment (rbharath) [inline deepchem/models/torch_models/dtnn.py:151]: Let's make this a public function and add an explicit unit test since it's quite complicated

Comment (rbharath) [reply deepchem/models/torch_models/dtnn.py:151]: This should document it's designed to work with CoulombMatrix outputs

Comment (rbharath) [reply deepchem/models/torch_models/dtnn.py:151]: For now, let's make a new file in utils for batching and add this there and document it

**Review by rbharath — COMMENTED** [review]
Getting close but a few more requests for cleanup/maintainability

File: deepchem/models/torch_models/dtnn.py (hunk 6 above)
Comment (GreatRSingh) [reply deepchem/models/torch_models/dtnn.py:17]: Done

File: deepchem/models/torch_models/dtnn.py (hunk 7 above)
Comment (GreatRSingh) [reply deepchem/models/torch_models/dtnn.py]: Done

File: deepchem/models/torch_models/dtnn.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
+from typing import List
+
+import torch.nn as nn
+import torch
+
+from deepchem.models.torch_models import layers
+
+
+class DTNN(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/dtnn.py:9]: This should be listed in docs/. Let's just add it to layers.rst.

**Review by rbharath — COMMENTED** [review]
Ok to merge in once added to docs

File: deepchem/models/torch_models/dtnn.py (hunk 8 above)
Comment (GreatRSingh) [reply deepchem/models/torch_models/dtnn.py:9]: Done

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**